File: code/sdks/sdk-ts/src/intents.ts
import type { Client } from "./client.js";
import type { components, paths } from "./generated/types.js";

/**
 * Helpers around POST /v1/intents and GET /v1/intents/{id}. The request /
 * response types are pulled from the OpenAPI-generated `components` so
 * adding a field to the server's zod schemas automatically tightens the
 * SDK signatures (T-12.3 derivation chain).
 */

export type AnyIntentInput = NonNullable<
  paths["/v1/intents"]["post"]["requestBody"]
>["content"]["application/json"];

export type IntentAccepted = components["schemas"]["IntentAccepted"];
export type IntentDetail = components["schemas"]["IntentDetail"];
export type IntentStatus = NonNullable<IntentAccepted["status"]>;

export const TERMINAL_STATUSES: readonly IntentStatus[] = ["confirmed", "failed"] as const;

export class IntentsApi {
  constructor(private readonly client: Client) {}

  async create(body: AnyIntentInput, opts?: { idempotencyKey?: string }): Promise<IntentAccepted> {
    return this.client.json<IntentAccepted>("POST", "/v1/intents", {
      body,
      headers: opts?.idempotencyKey ? { "idempotency-key": opts.idempotencyKey } : undefined,
    });
  }

  async get(intentId: string): Promise<IntentDetail> {
    return this.client.json<IntentDetail>("GET", `/v1/intents/${intentId}`);
  }

  /** Poll until the intent reaches a terminal status. Returns the final row. */
  async waitFor(
    intentId: string,
    opts?: { deadlineMs?: number; pollMs?: number },
  ): Promise<IntentDetail> {
    const deadline = Date.now() + (opts?.deadlineMs ?? 300_000);
    const pollMs = opts?.pollMs ?? 250;
    while (Date.now() < deadline) {
      const row = await this.get(intentId);
      if (row.status && TERMINAL_STATUSES.includes(row.status as IntentStatus)) return row;
      await new Promise((res) => setTimeout(res, pollMs));
    }
    throw new Error(`intent ${intentId} did not reach a terminal state in time`);
  }
}

File: code/sdks/sdk-py/src/zenfinance_sdk/intents.py
"""
Intent helpers (createIntent / getIntent / wait_for_intent).

Request and response shapes come from the generated pydantic models so
the SDK signatures stay in sync with the server's zod schemas (T-12.3).
"""

from __future__ import annotations

import time
from typing import Mapping

from .client import Client
from .generated.models import IntentAccepted, IntentDetail

TERMINAL_STATUSES = frozenset({"confirmed", "failed"})


class IntentsApi:
    def __init__(self, client: Client) -> None:
        self._client = client

    def create(
        self,
        body: Mapping[str, object],
        idempotency_key: str | None = None,
    ) -> IntentAccepted:
        headers: dict[str, str] = {}
        if idempotency_key:
            headers["idempotency-key"] = idempotency_key
        raw = self._client.request("POST", "/v1/intents", body=body, headers=headers)
        return IntentAccepted.model_validate(raw)

    def get(self, intent_id: str) -> IntentDetail:
        raw = self._client.request("GET", f"/v1/intents/{intent_id}")
        return IntentDetail.model_validate(raw)

    def wait_for(
        self,
        intent_id: str,
        deadline_s: float = 60.0,
        poll_s: float = 0.25,
    ) -> IntentDetail:
        deadline = time.monotonic() + deadline_s
        while time.monotonic() < deadline:
            row = self.get(intent_id)
            status = row.status.value if hasattr(row.status, "value") else row.status
            if status in TERMINAL_STATUSES:
                return row
            time.sleep(poll_s)
        raise RuntimeError(
            f"intent {intent_id} did not reach a terminal state within {deadline_s}s",
        )
